packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,445 @@
+            enumerable: true,
+            configurable: true,
+        },
+    });
+}
+
+function patchAttributesDuringUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instanceDefinition: Definition
+) {
+    // The below case patches observed attributes for the case where the HTML element is upgraded
+    // from a pre-existing one in the DOM.
+    const newObservedAttributes = getNewObservedAttributes(
+        registeredDefinition,
+        instanceDefinition
+    );
+    if (getNewObservedAttributes(registeredDefinition, instanceDefinition).size === 0) {
+        return;
+    }
+    const { attributeChangedCallback } = instanceDefinition;
+    // Approximate observedAttributes from the user class, but only for the new observed attributes
+    newObservedAttributes.forEach((name) => {
+        if (nativeHasAttribute.call(instance, name)) {
+            const newValue = nativeGetAttribute.call(instance, name);
+            attributeChangedCallback!.call(instance, name, null, newValue);
+        }
+    });
+}
+
+// User extends this HTMLElement, which returns the CE being upgraded
+let upgradingInstance: HTMLElement | undefined;
+
+// Helper to upgrade an instance with a CE definition using "constructor call trick"
+function internalUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instanceDefinition: Definition
+) {
+    setPrototypeOf(instance, instanceDefinition.UserCtor.prototype);
+    definitionForElement.set(instance, instanceDefinition);
+    // attributes patches when needed
+    if (instanceDefinition !== registeredDefinition) {
+        patchAttributes(instance, registeredDefinition, instanceDefinition);
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    // By `new`ing the UserCtor, we now jump to the constructor for the overridden global HTMLElement
+    // The reason this happens is that the UserCtor extends HTMLElement, so it calls the `super()`.
+    // Note that `upgradingInstance` is explicitly handled in the HTMLElement constructor.
+    new instanceDefinition.UserCtor();
+
+    patchAttributesDuringUpgrade(instance, registeredDefinition, instanceDefinition);
+}
+
+function getOrCreateDefinitionForConstructor(constructor: CustomElementConstructor): Definition {
+    if (!isFunction(constructor) || !isObject(constructor.prototype)) {
+        throw new TypeError('The referenced constructor is not a constructor.');
+    }
+    const definition = definitionForConstructor.get(constructor);
+    if (!isUndefined(definition)) {
+        return definition;
+    }
+    return createDefinitionRecord(constructor);
+}
+
+export function createScopedRegistry() {
```

Comment:
@caridy It occurred to me that we do not need `createScopedRegistry` at all. We can just directly export `defineScopedElement`.

From LWC's perspective, we only ever create one of these. And even if you call `createScopedRegistry` multiple times, it's actually still using the shared caches under the hood, so it's kind of pointless.

I'd like to think about how Locker could reuse this file, though, so let me know if that would be a bad idea from Locker's POV.

Reply:
Removed it.